### PR TITLE
"final" set of changes

### PR DESCRIPTION
**Pull request by emtempleton**
Still more to do after the class is over! But with this PR, I will add a script that loads and applies pre-trained models to a test set of documents, then compares those models against each other, and saves a figure summarizing those comparisons.

### CODE REVIEW

**Comment by emtempleton** [comment]
Note: Check out the updated README for an explicit list of all the improvements I made.

File: .travis.yml [hunk 1]
```diff
@@ -4,13 +4,16 @@ python:
...
+  - pip install codecov flake8 pytest bs4 ipython pandas nltk sklearn matplotlib
 
 script:
   - flake8 topic_modeling/scrape_training_data/*.py
   - pytest topic_modeling/scrape_training_data/*.py
   - flake8 topic_modeling/train_models/*.py
   - pytest topic_modeling/train_models/*.py
+  - flake8 topic_modeling/validate_models/*.py
+  - pytest topic_modeling/validate_models/*.py
+  - flake8 topic_modeling/*.py
```
Comment (yarikoptic) [inline .travis.yml:16]: ideally
- move all tests into test_*.y files under `tests/` submodules. And then call `pytest topic_modeling`
- enable coverage reporting
- also have just a single flake8 call.
For all of that consult e.g. shablona's .travis.yml: https://github.com/uwescience/shablona/blob/master/.travis.yml

Doing that in such more generic way would allow you to avoid modifying travis.yml, and also get an overall coverage for the entire topic_modeling.

File: README.md [hunk 2]
```diff
@@ -1,21 +1,47 @@
...
+This pipeline makes it easy to manipulate different parameters (number of topics, text pre-processing steps) and to use an external "test" to pick the best performing model. This "winning" model can then be applied to documents of interest.
 
-2. The purpose of the model validation step is to compare models that vary across a set of parameters. I will create functions that allow the user to set these parameters (number of topics, text preprocessing steps, etc).
+# Steps
 
-3. Currently, I create figures that summarize the results of model comparisons in a separate script. I would like these figures to run and output by default. I think I separated this initially because I was having issues creating figures in the cluster. I’ll add a parameter that will allow the user to turn off this option if they want (e.g., if it avoids issues on a cluster).
+Each step of this process is contained in a separate directory.
```
Comment (yarikoptic) [inline README.md:13]: Thought to suggest to reword but then mentioned that those (e.g. `topic_modeling/validate_models`) are indeed just subdirectories, not submodules (may be that is why you did pytest separately for them). Add `__init__.py`, if even just a blank one to each one of them, and then they will magically turn into submodules so you could e.g. `from topic_modeling import validate_models` and state here that they are submodules.

Comment (yarikoptic) [reply README.md:13]: then in the longer run I would recommend interface those "scripts" as subcommands to the ultimate `topic_modeling` command, e.g. `topic_modeling scrape_training_data`.  You could achieve that with https://docs.python.org/3/library/argparse.html  but the easiest way would be to use something like https://click.palletsprojects.com module . 
That would allow to just have your `topic_modeling` package installed, without polluting the space of commands in cmdline, while making them all nicely groupped and with nice `--help`

File: README.md [hunk 3]
```diff
@@ -1,21 +1,47 @@
...
-6. Some of my paths are hard-coded. I will be more generalizable about calling path names.
+Contained in this directory are scripts that find, scrape, and store text from articles in The Dartmouth daily newspaer archives. 
 
-7. I will make these scripts PEP8 compatible.   
\ No newline at end of file
+## 2. Train Models
+
+./train_models
+
+The script 'train_models.py' takes the training data created in Step 1, pre-processes it, and uses it to train a series of topic models that vary across a set of parameters (number of topics, text pre-processing steps). These models are pickled and stored in 'models_pickles'. Each model also has a corresponding text files in 'models_topics' that lists the top words associated with topic. These collections of words allow the user to provide model 'labels' and to quickly assess the interpretability of the topics.
```
Comment (yarikoptic) [inline README.md:25]: you might better just call that directory `trained_models` or alike, and just rely how they are stored in the extension of the file, which you already have `.pkl`.  you might later follow scikit-learn's advice and use joblib's functionality to store/load them https://scikit-learn.org/stable/modules/model_persistence.html and store into `.joblib` files.  If you keep working with only one type of the model, creating .json (or some other e.g. .yaml) serialization might help with maintainability, compatibility (pickling could cause troubles across versions of python or even possibly scikit-learn itself) and ability to "see" the model. See e.g. approach taken at https://stackabuse.com/scikit-learn-save-and-restore-models/ which might be a bit more appropriate/lightweight here in comparison to the one you found in https://cmry.github.io/notes/serialize

File: README.md [hunk 4]
```diff
@@ -1,21 +1,47 @@
...
+6. Scripted the model comparision step (I was previously creating these correlation matrices and figures in an unruly jupyter notebook).
+7. Fixed the issue of sentences being smooshed together when I scraped the articles.
+8. Allowed for an important text preprocessing step (stemming) to be set as a parameter.
+9. Finally stored trained models. I was previously re-running them at every step which was a huge waste of resources. I stored the models by pickling them, though I know Yarik has a preference for .json files. However, JSON serialization from sklearn pipelines is not straightforward (see one article here: https://cmry.github.io/notes/serialize) so I've left it as a pickle for now. 
+10. No longer hardcoded paths (though currently relying on my specific file structure. will want to allow for users to set their own top level directories eventually).
+11. Meta point: Actually worked with git! Now comfortable creating branches, commiting changes, submitting PRs, closing issues, and merging!
```
Comment (yarikoptic) [inline README.md:47]: you indeed did great job arranging your PRs for the reviews, thank you for that! ;)

File: topic_modeling/scrape_training_data/get_article_links.py [hunk 5]
```diff
@@ -5,50 +5,78 @@
...
-            if link not in article_links:
-                if 'article' in link.split('/'):
-                    article_links.append(link)
+        for page in pages:
 
+            html_page = urllib.request.urlopen(
+                "http://www.thedartmouth.com/search?q=0&page={}&"
+                "ti=0&tg=0&ty=0&ts_month=0&ts_day=0&ts_year={}&"
+                "te_month=0&te_day=0&te_year={}&s=0&au=0&o=0&"
+                "a=1".format(page, year_url, year_url+1))
```
Comment (yarikoptic) [inline topic_modeling/scrape_training_data/get_article_links.py:52]: with such long strings, best to rely not on positional arguments for the `format()` but specify them as keyword arguments and use then `{page}`, `{year}`, `{next_year}` - so that url would be easy to read and thus avoid possible mistakes when you have to edit it

File: topic_modeling/scrape_training_data/get_article_links.py [hunk 6]
```diff
@@ -5,50 +5,78 @@
...
-requests = 0
+    Parameters
+    ----------
+    start_year : int
+        Year in the archives to start looking for articles
+    end_year : int
+        Year in the archives to stop looking for articles
+    num_pages : int
+        Number of archive pages that fit each year range to
+        inspect
```
Comment (yarikoptic) [inline topic_modeling/scrape_training_data/get_article_links.py:22]: isn't this something what varies from year to year and should be deduced from the pages themselves?

File: topic_modeling/scrape_training_data/get_article_links.py (hunk 5 above)
Comment (yarikoptic) [reply topic_modeling/scrape_training_data/get_article_links.py:52]: eventually, if you need to scrape some other bigger site, or even now if it takes too long, you might like to learn/use https://scrapy.org which could crawl in parallel etc

File: topic_modeling/scrape_training_data/scrape_articles.py [hunk 7]
```diff
@@ -53,52 +53,72 @@ def test_log_error():
...
+        article is dumped into that text file. The text file is named
+        with the same naming convention The Dartmouth uses for their
+        articles URLs.
+
+    """
+
+    # need two different ways to set directory to pass
+    # Travis CI (because .travis file is called from
+    # different directory)
+    current_file = os.getcwd().split('/')[-1]
```
Comment (yarikoptic) [inline topic_modeling/scrape_training_data/scrape_articles.py:74]: there is `os.path.split` which is recommended to be used over string's split since it would then work also on exotic systems such as windows

**Review by yarikoptic — COMMENTED** [review]
very nice.  I left comments about specific aspects I caught.
My only major grief is really sparse amount of testing.  You could start by adding basic smoke tests of your functions - e.g. reloading those pickled models, and saving them again, loading again and verifying that they are identical.  Would require only a few lines of code but would at least guarantee that you can still load/store the models, possibly correctly.  Ideally creating some really lightweight test case where you know the ground truth and going through all of your functionality and verifying that you are obtaining expected results.
